File: src/main/java/com/github/quiltservertools/interdimensional/world/RuntimeWorldManager.java
package com.github.quiltservertools.interdimensional.world;

import com.github.quiltservertools.interdimensional.Interdimensional;
import com.google.gson.JsonObject;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.util.Identifier;
import xyz.nucleoid.fantasy.Fantasy;
import xyz.nucleoid.fantasy.RuntimeWorldConfig;
import xyz.nucleoid.fantasy.RuntimeWorldHandle;

import java.util.ArrayList;
import java.util.List;

public class RuntimeWorldManager {
    private static final List<RuntimeWorldHandle> runtimeDimensionHandlers = new ArrayList<>();

    public static void add(RuntimeWorldConfig config, Identifier identifier) {
        RuntimeWorldHandle handle = Interdimensional.FANTASY.getOrOpenPersistentWorld(identifier, config);
        runtimeDimensionHandlers.add(handle);
    }

    public static void remove(RuntimeWorldHandle handle) {
        runtimeDimensionHandlers.remove(handle);
        Interdimensional.LOGGER.info("Removed dimension" + handle.asWorld().getRegistryKey().getValue());
        handle.delete();
    }

    public static List<JsonObject> closeAll() {
        var list = new ArrayList<JsonObject>();
        runtimeDimensionHandlers.forEach(handle -> {
            var object = new JsonObject();
            object.addProperty("identifier", handle.asWorld().getRegistryKey().getValue().toString());
            list.add(object);
        });
        return list;
    }

    public static ServerWorld get(Identifier identifier, MinecraftServer server) {
        var result = runtimeDimensionHandlers.stream().filter(h -> h.asWorld().getRegistryKey().getValue().equals(identifier)).findFirst();
        if (result.isPresent()) {
            return result.get().asWorld();
        } else {
            return server.getOverworld();
        }
    }

    public static RuntimeWorldHandle getHandle(Identifier identifier, MinecraftServer server) {
        var result = runtimeDimensionHandlers.stream().filter(h -> h.asWorld().getRegistryKey().getValue().equals(identifier)).findFirst();
        return result.orElse(null);
    }
}

File: src/main/kotlin/net/quiltservertools/interdimensional/world/RuntimeWorldManager.kt
package net.quiltservertools.interdimensional.world

import net.quiltservertools.interdimensional.Interdimensional
import com.google.gson.JsonObject
import net.minecraft.server.MinecraftServer
import net.minecraft.server.world.ServerWorld
import net.minecraft.util.Identifier
import xyz.nucleoid.fantasy.RuntimeWorldConfig
import xyz.nucleoid.fantasy.RuntimeWorldHandle
import java.util.function.Consumer

object RuntimeWorldManager {
    private val runtimeDimensionHandlers: MutableList<RuntimeWorldHandle> = ArrayList()
    fun add(config: RuntimeWorldConfig, identifier: Identifier) {
        val handle = Interdimensional.FANTASY.getOrOpenPersistentWorld(identifier, config)
        runtimeDimensionHandlers.add(handle)
    }

    fun remove(handle: RuntimeWorldHandle) {
        runtimeDimensionHandlers.remove(handle)
        Interdimensional.LOGGER.info("Removed dimension${handle.asWorld().registryKey.value}")
        handle.delete()
    }

    fun closeAll(): List<JsonObject> {
        val list = ArrayList<JsonObject>()
        runtimeDimensionHandlers.forEach(Consumer { handle: RuntimeWorldHandle ->
            val `object` = JsonObject()
            `object`.addProperty("identifier", handle.asWorld().registryKey.value.toString())
            list.add(`object`)
        })
        return list
    }

    operator fun get(identifier: Identifier, server: MinecraftServer): ServerWorld {
        val result = runtimeDimensionHandlers.stream().filter { h: RuntimeWorldHandle ->
            h.asWorld().registryKey.value == identifier
        }.findFirst()
        return if (result.isPresent) {
            result.get().asWorld()
        } else {
            server.overworld
        }
    }

    fun getHandle(identifier: Identifier): RuntimeWorldHandle {
        val result = runtimeDimensionHandlers.first { h: RuntimeWorldHandle ->
            h.asWorld().registryKey.value == identifier
        }
        return result
    }
}